##### Copyright 2019 The TensorFlow Authors.

In [1]:
#@title Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

In [2]:
#@title MIT License
#
# Copyright (c) 2017 François Chollet
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
# DEALINGS IN THE SOFTWARE.

# Text classification with TensorFlow Hub: Movie reviews

<table class="tfo-notebook-buttons" align="left">
  <td>
    <a target="_blank" href="https://www.tensorflow.org/tutorials/keras/text_classification_with_hub"><img src="https://www.tensorflow.org/images/tf_logo_32px.png" />View on TensorFlow.org</a>
  </td>
  <td>
    <a target="_blank" href="https://colab.research.google.com/github/tensorflow/docs/blob/master/site/en/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://www.tensorflow.org/images/colab_logo_32px.png" />Run in Google Colab</a>
  </td>
  <td>
    <a target="_blank" href="https://github.com/tensorflow/docs/blob/master/site/en/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://www.tensorflow.org/images/GitHub-Mark-32px.png" />View source on GitHub</a>
  </td>
  <td>
    <a href="https://storage.googleapis.com/tensorflow_docs/docs/site/en/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://www.tensorflow.org/images/download_logo_32px.png" />Download notebook</a>
  </td>
</table>

This notebook classifies movie reviews as *positive* or *negative* using the text of the review. This is an example of *binary*—or two-class—classification, an important and widely applicable kind of machine learning problem.

The tutorial demonstrates the basic application of transfer learning with TensorFlow Hub and Keras.

We'll use the [IMDB dataset](https://www.tensorflow.org/api_docs/python/tf/keras/datasets/imdb) that contains the text of 50,000 movie reviews from the [Internet Movie Database](https://www.imdb.com/). These are split into 25,000 reviews for training and 25,000 reviews for testing. The training and testing sets are *balanced*, meaning they contain an equal number of positive and negative reviews. 

This notebook uses [tf.keras](https://www.tensorflow.org/guide/keras), a high-level API to build and train models in TensorFlow, and [TensorFlow Hub](https://www.tensorflow.org/hub), a library and platform for transfer learning. For a more advanced text classification tutorial using `tf.keras`, see the [MLCC Text Classification Guide](https://developers.google.com/machine-learning/guides/text-classification/).

In [3]:
import numpy as np

import tensorflow as tf

!pip install -q tensorflow-hub
!pip install -q tfds-nightly
import tensorflow_hub as hub
import tensorflow_datasets as tfds

print("Version: ", tf.__version__)
print("Eager mode: ", tf.executing_eagerly())
print("Hub version: ", hub.__version__)
print("GPU is", "available" if tf.config.experimental.list_physical_devices("GPU") else "NOT AVAILABLE")

Version:  2.1.0
Eager mode:  True
Hub version:  0.7.0
GPU is available


## Download the IMDB dataset

The IMDB dataset is available on [imdb reviews](https://www.tensorflow.org/datasets/catalog/imdb_reviews) or on [TensorFlow datasets](https://www.tensorflow.org/datasets). The following code downloads the IMDB dataset to your machine (or the colab runtime):

In [4]:
# Split the training set into 60% and 40%, so we'll end up with 15,000 examples
# for training, 10,000 examples for validation and 25,000 examples for testing.
train_data, validation_data, test_data = tfds.load(
    name="imdb_reviews", 
    split=('train[:60%]', 'train[60%:]', 'test'),
    as_supervised=True)

Shuffling and writing examples to /home/kbuilder/tensorflow_datasets/imdb_reviews/plain_text/1.0.0.incomplete2L9LMH/imdb_reviews-train.tfrecord


Shuffling and writing examples to /home/kbuilder/tensorflow_datasets/imdb_reviews/plain_text/1.0.0.incomplete2L9LMH/imdb_reviews-test.tfrecord


Shuffling and writing examples to /home/kbuilder/tensorflow_datasets/imdb_reviews/plain_text/1.0.0.incomplete2L9LMH/imdb_reviews-unsupervised.tfrecord


Dataset imdb_reviews downloaded and prepared to /home/kbuilder/tensorflow_datasets/imdb_reviews/plain_text/1.0.0. Subsequent calls will reuse this data.


## Explore the data 

Let's take a moment to understand the format of the data. Each example is a sentence representing the movie review and a corresponding label. The sentence is not preprocessed in any way. The label is an integer value of either 0 or 1, where 0 is a negative review, and 1 is a positive review.

Let's print first 10 examples.

In [5]:
train_examples_batch, train_labels_batch = next(iter(train_data.batch(10)))
train_examples_batch

<tf.Tensor: shape=(10,), dtype=string, numpy=
array([b"This was an absolutely terrible movie. Don't be lured in by Christopher Walken or Michael Ironside. Both are great actors, but this must simply be their worst role in history. Even their great acting could not redeem this movie's ridiculous storyline. This movie is an early nineties US propaganda piece. The most pathetic scenes were those when the Columbian rebels were making their cases for revolutions. Maria Conchita Alonso appeared phony, and her pseudo-love affair with Walken was nothing but a pathetic emotional plug in a movie that was devoid of any real meaning. I am disappointed that there are movies like this, ruining actor's like Christopher Walken's good name. I could barely sit through it.",
       b'I have been known to fall asleep during films, but this is usually due to a combination of things including, really tired, being warm and comfortable on the sette and having just eaten a lot. However on this occasion I fell 

Let's also print the first 10 labels.

In [6]:
train_labels_batch

<tf.Tensor: shape=(10,), dtype=int64, numpy=array([0, 0, 0, 1, 1, 1, 0, 0, 0, 0])>

## Build the model

The neural network is created by stacking layers—this requires three main architectural decisions:

* How to represent the text?
* How many layers to use in the model?
* How many *hidden units* to use for each layer?

In this example, the input data consists of sentences. The labels to predict are either 0 or 1.

One way to represent the text is to convert sentences into embeddings vectors. We can use a pre-trained text embedding as the first layer, which will have three advantages:

*   we don't have to worry about text preprocessing,
*   we can benefit from transfer learning,
*   the embedding has a fixed size, so it's simpler to process.

For this example we will use a **pre-trained text embedding model** from [TensorFlow Hub](https://www.tensorflow.org/hub) called [google/tf2-preview/gnews-swivel-20dim/1](https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim/1).

There are three other pre-trained models to test for the sake of this tutorial:

* [google/tf2-preview/gnews-swivel-20dim-with-oov/1](https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim-with-oov/1) - same as [google/tf2-preview/gnews-swivel-20dim/1](https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim/1), but with 2.5% vocabulary converted to OOV buckets. This can help if vocabulary of the task and vocabulary of the model don't fully overlap.
* [google/tf2-preview/nnlm-en-dim50/1](https://tfhub.dev/google/tf2-preview/nnlm-en-dim50/1) - A much larger model with ~1M vocabulary size and 50 dimensions.
* [google/tf2-preview/nnlm-en-dim128/1](https://tfhub.dev/google/tf2-preview/nnlm-en-dim128/1) - Even larger model with ~1M vocabulary size and 128 dimensions.

Let's first create a Keras layer that uses a TensorFlow Hub model to embed the sentences, and try it out on a couple of input examples. Note that no matter the length of the input text, the output shape of the embeddings is: `(num_examples, embedding_dimension)`.

In [7]:
embedding = "https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim/1"
hub_layer = hub.KerasLayer(embedding, input_shape=[], 
                           dtype=tf.string, trainable=True)
hub_layer(train_examples_batch[:3])

<tf.Tensor: shape=(3, 20), dtype=float32, numpy=
array([[ 1.765786  , -3.882232  ,  3.9134233 , -1.5557289 , -3.3362343 ,
        -1.7357955 , -1.9954445 ,  1.2989551 ,  5.081598  , -1.1041286 ,
        -2.0503852 , -0.72675157, -0.65675956,  0.24436149, -3.7208383 ,
         2.0954835 ,  2.2969332 , -2.0689783 , -2.9489717 , -1.1315987 ],
       [ 1.8804485 , -2.5852382 ,  3.4066997 ,  1.0982676 , -4.056685  ,
        -4.891284  , -2.785554  ,  1.3874227 ,  3.8476458 , -0.9256538 ,
        -1.896706  ,  1.2113281 ,  0.11474707,  0.76209456, -4.8791065 ,
         2.906149  ,  4.7087674 , -2.3652055 , -3.5015898 , -1.6390051 ],
       [ 0.71152234, -0.6353217 ,  1.7385626 , -1.1168286 , -0.5451594 ,
        -1.1808156 ,  0.09504455,  1.4653089 ,  0.66059524,  0.79308075,
        -2.2268345 ,  0.07446612, -1.4075904 , -0.70645386, -1.907037  ,
         1.4419787 ,  1.9551861 , -0.42660055, -2.8022065 ,  0.43727064]],
      dtype=float32)>

Let's now build the full model:

In [8]:
model = tf.keras.Sequential()
model.add(hub_layer)
model.add(tf.keras.layers.Dense(16, activation='relu'))
model.add(tf.keras.layers.Dense(1))

model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
keras_layer (KerasLayer)     (None, 20)                400020    
_________________________________________________________________
dense (Dense)                (None, 16)                336       
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 17        
Total params: 400,373
Trainable params: 400,373
Non-trainable params: 0
_________________________________________________________________


The layers are stacked sequentially to build the classifier:

1. The first layer is a TensorFlow Hub layer. This layer uses a pre-trained Saved Model to map a sentence into its embedding vector. The pre-trained text embedding model that we are using ([google/tf2-preview/gnews-swivel-20dim/1](https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim/1)) splits the sentence into tokens, embeds each token and then combines the embedding. The resulting dimensions are: `(num_examples, embedding_dimension)`.
2. This fixed-length output vector is piped through a fully-connected (`Dense`) layer with 16 hidden units.
3. The last layer is densely connected with a single output node. Using the `sigmoid` activation function, this value is a float between 0 and 1, representing a probability, or confidence level.

Let's compile the model.

### Loss function and optimizer

A model needs a loss function and an optimizer for training. Since this is a binary classification problem and the model outputs a probability (a single-unit layer with a sigmoid activation), we'll use the `binary_crossentropy` loss function. 

This isn't the only choice for a loss function, you could, for instance, choose `mean_squared_error`. But, generally, `binary_crossentropy` is better for dealing with probabilities—it measures the "distance" between probability distributions, or in our case, between the ground-truth distribution and the predictions.

Later, when we are exploring regression problems (say, to predict the price of a house), we will see how to use another loss function called mean squared error.

Now, configure the model to use an optimizer and a loss function:

In [9]:
model.compile(optimizer='adam',
              loss=tf.keras.losses.BinaryCrossentropy(from_logits=True),
              metrics=['accuracy'])

## Train the model

Train the model for 20 epochs in mini-batches of 512 samples. This is 20 iterations over all samples in the `x_train` and `y_train` tensors. While training, monitor the model's loss and accuracy on the 10,000 samples from the validation set:

In [10]:
history = model.fit(train_data.shuffle(10000).batch(512),
                    epochs=20,
                    validation_data=validation_data.batch(512),
                    verbose=1)

Epoch 1/20


      1/Unknown - 1s 940ms/step - loss: 0.8515 - accuracy: 0.4629

      2/Unknown - 1s 497ms/step - loss: 0.8342 - accuracy: 0.4775

      3/Unknown - 1s 348ms/step - loss: 0.8058 - accuracy: 0.4883

      4/Unknown - 1s 274ms/step - loss: 0.7984 - accuracy: 0.4927

      5/Unknown - 1s 229ms/step - loss: 0.8005 - accuracy: 0.4949

      6/Unknown - 1s 198ms/step - loss: 0.7971 - accuracy: 0.5016

      7/Unknown - 1s 177ms/step - loss: 0.7960 - accuracy: 0.5042

      8/Unknown - 1s 160ms/step - loss: 0.7922 - accuracy: 0.5078

      9/Unknown - 1s 148ms/step - loss: 0.7875 - accuracy: 0.5117

     10/Unknown - 1s 138ms/step - loss: 0.7827 - accuracy: 0.5158

     11/Unknown - 1s 129ms/step - loss: 0.7766 - accuracy: 0.5210

     12/Unknown - 1s 122ms/step - loss: 0.7737 - accuracy: 0.5215

     13/Unknown - 2s 117ms/step - loss: 0.7710 - accuracy: 0.5234

     14/Unknown - 2s 112ms/step - loss: 0.7692 - accuracy: 0.5227

     15/Unknown - 2s 107ms/step - loss: 0.7670 - accuracy: 0.5233

     16/Unknown - 2s 103ms/step - loss: 0.7640 - accuracy: 0.5260

     17/Unknown - 2s 100ms/step - loss: 0.7619 - accuracy: 0.5277

     18/Unknown - 2s 96ms/step - loss: 0.7601 - accuracy: 0.5285 

     19/Unknown - 2s 94ms/step - loss: 0.7572 - accuracy: 0.5303

     20/Unknown - 2s 91ms/step - loss: 0.7549 - accuracy: 0.5322

     21/Unknown - 2s 89ms/step - loss: 0.7528 - accuracy: 0.5325

     22/Unknown - 2s 87ms/step - loss: 0.7517 - accuracy: 0.5344

     23/Unknown - 2s 85ms/step - loss: 0.7496 - accuracy: 0.5363

     24/Unknown - 2s 84ms/step - loss: 0.7473 - accuracy: 0.5387

     25/Unknown - 2s 82ms/step - loss: 0.7443 - accuracy: 0.5405

     26/Unknown - 2s 81ms/step - loss: 0.7416 - accuracy: 0.5430

     27/Unknown - 2s 79ms/step - loss: 0.7403 - accuracy: 0.5442

     28/Unknown - 2s 78ms/step - loss: 0.7393 - accuracy: 0.5460

     29/Unknown - 2s 77ms/step - loss: 0.7371 - accuracy: 0.5479

     30/Unknown - 2s 75ms/step - loss: 0.7369 - accuracy: 0.5479

30/30 [==============================] - 3s 115ms/step - loss: 0.7369 - accuracy: 0.5479 - val_loss: 0.6822 - val_accuracy: 0.5842


Epoch 2/20


 1/30 [>.............................] - ETA: 9s - loss: 0.6772 - accuracy: 0.5898

 2/30 [=>............................] - ETA: 5s - loss: 0.6692 - accuracy: 0.6055

 3/30 [==>...........................] - ETA: 4s - loss: 0.6744 - accuracy: 0.5924

 5/30 [====>.........................] - ETA: 2s - loss: 0.6698 - accuracy: 0.6000

 7/30 [======>.......................] - ETA: 2s - loss: 0.6675 - accuracy: 0.6004

 8/30 [=======>......................] - ETA: 1s - loss: 0.6693 - accuracy: 0.5969

 9/30 [========>.....................] - ETA: 1s - loss: 0.6655 - accuracy: 0.6035

11/30 [==========>...................] - ETA: 1s - loss: 0.6681 - accuracy: 0.6044

13/30 [============>.................] - ETA: 1s - loss: 0.6663 - accuracy: 0.6028

15/30 [==============>...............] - ETA: 1s - loss: 0.6642 - accuracy: 0.6052

17/30 [================>.............] - ETA: 0s - loss: 0.6623 - accuracy: 0.6062

19/30 [==================>...........] - ETA: 0s - loss: 0.6584 - accuracy: 0.6104

21/30 [====================>.........] - ETA: 0s - loss: 0.6554 - accuracy: 0.6105

23/30 [======================>.......] - ETA: 0s - loss: 0.6531 - accuracy: 0.6110

25/30 [========================>.....] - ETA: 0s - loss: 0.6505 - accuracy: 0.6128

27/30 [==========================>...] - ETA: 0s - loss: 0.6491 - accuracy: 0.6157

29/30 [============================>.] - ETA: 0s - loss: 0.6463 - accuracy: 0.6177

30/30 [==============================] - 3s 92ms/step - loss: 0.6454 - accuracy: 0.6176 - val_loss: 0.6233 - val_accuracy: 0.6409


Epoch 3/20


 1/30 [>.............................] - ETA: 9s - loss: 0.6100 - accuracy: 0.6523

 2/30 [=>............................] - ETA: 5s - loss: 0.6049 - accuracy: 0.6611

 3/30 [==>...........................] - ETA: 4s - loss: 0.6027 - accuracy: 0.6608

 4/30 [===>..........................] - ETA: 3s - loss: 0.6054 - accuracy: 0.6592

 6/30 [=====>........................] - ETA: 2s - loss: 0.6072 - accuracy: 0.6556

 8/30 [=======>......................] - ETA: 1s - loss: 0.6068 - accuracy: 0.6541

 9/30 [========>.....................] - ETA: 1s - loss: 0.6055 - accuracy: 0.6552

11/30 [==========>...................] - ETA: 1s - loss: 0.6009 - accuracy: 0.6587

13/30 [============>.................] - ETA: 1s - loss: 0.5982 - accuracy: 0.6594

15/30 [==============>...............] - ETA: 1s - loss: 0.5978 - accuracy: 0.6582

17/30 [================>.............] - ETA: 0s - loss: 0.5983 - accuracy: 0.6574

19/30 [==================>...........] - ETA: 0s - loss: 0.5978 - accuracy: 0.6567

21/30 [====================>.........] - ETA: 0s - loss: 0.5960 - accuracy: 0.6601

23/30 [======================>.......] - ETA: 0s - loss: 0.5941 - accuracy: 0.6610

25/30 [========================>.....] - ETA: 0s - loss: 0.5927 - accuracy: 0.6616

27/30 [==========================>...] - ETA: 0s - loss: 0.5918 - accuracy: 0.6618

29/30 [============================>.] - ETA: 0s - loss: 0.5895 - accuracy: 0.6659

30/30 [==============================] - 3s 93ms/step - loss: 0.5875 - accuracy: 0.6673 - val_loss: 0.5719 - val_accuracy: 0.6772


Epoch 4/20


 1/30 [>.............................] - ETA: 9s - loss: 0.5647 - accuracy: 0.6875

 2/30 [=>............................] - ETA: 5s - loss: 0.5581 - accuracy: 0.6768

 3/30 [==>...........................] - ETA: 4s - loss: 0.5645 - accuracy: 0.6829

 5/30 [====>.........................] - ETA: 2s - loss: 0.5598 - accuracy: 0.6805

 7/30 [======>.......................] - ETA: 2s - loss: 0.5566 - accuracy: 0.6825

 8/30 [=======>......................] - ETA: 1s - loss: 0.5536 - accuracy: 0.6858

 9/30 [========>.....................] - ETA: 1s - loss: 0.5518 - accuracy: 0.6882

11/30 [==========>...................] - ETA: 1s - loss: 0.5502 - accuracy: 0.6932

13/30 [============>.................] - ETA: 1s - loss: 0.5524 - accuracy: 0.6931

15/30 [==============>...............] - ETA: 1s - loss: 0.5504 - accuracy: 0.6965

17/30 [================>.............] - ETA: 0s - loss: 0.5473 - accuracy: 0.7001

19/30 [==================>...........] - ETA: 0s - loss: 0.5444 - accuracy: 0.7016

21/30 [====================>.........] - ETA: 0s - loss: 0.5419 - accuracy: 0.7030

23/30 [======================>.......] - ETA: 0s - loss: 0.5420 - accuracy: 0.7035

25/30 [========================>.....] - ETA: 0s - loss: 0.5396 - accuracy: 0.7077

27/30 [==========================>...] - ETA: 0s - loss: 0.5393 - accuracy: 0.7075

29/30 [============================>.] - ETA: 0s - loss: 0.5376 - accuracy: 0.7092

30/30 [==============================] - 3s 91ms/step - loss: 0.5358 - accuracy: 0.7096 - val_loss: 0.5262 - val_accuracy: 0.7155


Epoch 5/20


 1/30 [>.............................] - ETA: 9s - loss: 0.4915 - accuracy: 0.7578

 2/30 [=>............................] - ETA: 5s - loss: 0.4919 - accuracy: 0.7275

 3/30 [==>...........................] - ETA: 4s - loss: 0.4917 - accuracy: 0.7402

 4/30 [===>..........................] - ETA: 3s - loss: 0.4980 - accuracy: 0.7402

 6/30 [=====>........................] - ETA: 2s - loss: 0.4999 - accuracy: 0.7383

 8/30 [=======>......................] - ETA: 1s - loss: 0.5010 - accuracy: 0.7390

 9/30 [========>.....................] - ETA: 1s - loss: 0.5004 - accuracy: 0.7383

11/30 [==========>...................] - ETA: 1s - loss: 0.5023 - accuracy: 0.7369

13/30 [============>.................] - ETA: 1s - loss: 0.4994 - accuracy: 0.7378

15/30 [==============>...............] - ETA: 1s - loss: 0.4988 - accuracy: 0.7379

17/30 [================>.............] - ETA: 0s - loss: 0.4964 - accuracy: 0.7408

19/30 [==================>...........] - ETA: 0s - loss: 0.4980 - accuracy: 0.7423

21/30 [====================>.........] - ETA: 0s - loss: 0.4959 - accuracy: 0.7448

23/30 [======================>.......] - ETA: 0s - loss: 0.4939 - accuracy: 0.7466

25/30 [========================>.....] - ETA: 0s - loss: 0.4920 - accuracy: 0.7475

27/30 [==========================>...] - ETA: 0s - loss: 0.4898 - accuracy: 0.7497

29/30 [============================>.] - ETA: 0s - loss: 0.4893 - accuracy: 0.7503

30/30 [==============================] - 3s 91ms/step - loss: 0.4871 - accuracy: 0.7508 - val_loss: 0.4842 - val_accuracy: 0.7512


Epoch 6/20


 1/30 [>.............................] - ETA: 9s - loss: 0.4638 - accuracy: 0.7812

 2/30 [=>............................] - ETA: 5s - loss: 0.4578 - accuracy: 0.7822

 3/30 [==>...........................] - ETA: 4s - loss: 0.4611 - accuracy: 0.7760

 5/30 [====>.........................] - ETA: 2s - loss: 0.4597 - accuracy: 0.7875

 7/30 [======>.......................] - ETA: 2s - loss: 0.4625 - accuracy: 0.7801

 8/30 [=======>......................] - ETA: 1s - loss: 0.4595 - accuracy: 0.7793

 9/30 [========>.....................] - ETA: 1s - loss: 0.4603 - accuracy: 0.7804

11/30 [==========>...................] - ETA: 1s - loss: 0.4581 - accuracy: 0.7795

12/30 [===========>..................] - ETA: 1s - loss: 0.4568 - accuracy: 0.7799

14/30 [=============>................] - ETA: 1s - loss: 0.4522 - accuracy: 0.7808

16/30 [===============>..............] - ETA: 0s - loss: 0.4507 - accuracy: 0.7830

18/30 [=================>............] - ETA: 0s - loss: 0.4508 - accuracy: 0.7817

20/30 [===================>..........] - ETA: 0s - loss: 0.4497 - accuracy: 0.7817

22/30 [=====================>........] - ETA: 0s - loss: 0.4467 - accuracy: 0.7839

24/30 [=======================>......] - ETA: 0s - loss: 0.4464 - accuracy: 0.7837

26/30 [=========================>....] - ETA: 0s - loss: 0.4439 - accuracy: 0.7863

28/30 [===========================>..] - ETA: 0s - loss: 0.4438 - accuracy: 0.7845

30/30 [==============================] - 3s 94ms/step - loss: 0.4420 - accuracy: 0.7853 - val_loss: 0.4461 - val_accuracy: 0.7847


Epoch 7/20


 1/30 [>.............................] - ETA: 10s - loss: 0.4146 - accuracy: 0.7969

 2/30 [=>............................] - ETA: 5s - loss: 0.4232 - accuracy: 0.8008 

 3/30 [==>...........................] - ETA: 4s - loss: 0.4117 - accuracy: 0.8158

 5/30 [====>.........................] - ETA: 2s - loss: 0.4142 - accuracy: 0.8125

 7/30 [======>.......................] - ETA: 2s - loss: 0.4071 - accuracy: 0.8145

 8/30 [=======>......................] - ETA: 1s - loss: 0.4105 - accuracy: 0.8069

 9/30 [========>.....................] - ETA: 1s - loss: 0.4122 - accuracy: 0.8030

11/30 [==========>...................] - ETA: 1s - loss: 0.4108 - accuracy: 0.8036

13/30 [============>.................] - ETA: 1s - loss: 0.4101 - accuracy: 0.8059

15/30 [==============>...............] - ETA: 1s - loss: 0.4074 - accuracy: 0.8092

17/30 [================>.............] - ETA: 0s - loss: 0.4072 - accuracy: 0.8088

19/30 [==================>...........] - ETA: 0s - loss: 0.4061 - accuracy: 0.8109

21/30 [====================>.........] - ETA: 0s - loss: 0.4040 - accuracy: 0.8120

23/30 [======================>.......] - ETA: 0s - loss: 0.4029 - accuracy: 0.8117

25/30 [========================>.....] - ETA: 0s - loss: 0.4028 - accuracy: 0.8112

27/30 [==========================>...] - ETA: 0s - loss: 0.4024 - accuracy: 0.8116

29/30 [============================>.] - ETA: 0s - loss: 0.4017 - accuracy: 0.8134

30/30 [==============================] - 3s 92ms/step - loss: 0.4011 - accuracy: 0.8139 - val_loss: 0.4137 - val_accuracy: 0.8028


Epoch 8/20


 1/30 [>.............................] - ETA: 9s - loss: 0.3407 - accuracy: 0.8652

 2/30 [=>............................] - ETA: 5s - loss: 0.3748 - accuracy: 0.8398

 3/30 [==>...........................] - ETA: 4s - loss: 0.3763 - accuracy: 0.8411

 5/30 [====>.........................] - ETA: 2s - loss: 0.3796 - accuracy: 0.8238

 7/30 [======>.......................] - ETA: 2s - loss: 0.3712 - accuracy: 0.8315

 8/30 [=======>......................] - ETA: 1s - loss: 0.3745 - accuracy: 0.8337

10/30 [=========>....................] - ETA: 1s - loss: 0.3743 - accuracy: 0.8355

12/30 [===========>..................] - ETA: 1s - loss: 0.3721 - accuracy: 0.8377

14/30 [=============>................] - ETA: 1s - loss: 0.3721 - accuracy: 0.8365

16/30 [===============>..............] - ETA: 0s - loss: 0.3712 - accuracy: 0.8337

18/30 [=================>............] - ETA: 0s - loss: 0.3737 - accuracy: 0.8291

20/30 [===================>..........] - ETA: 0s - loss: 0.3712 - accuracy: 0.8316

22/30 [=====================>........] - ETA: 0s - loss: 0.3690 - accuracy: 0.8350

24/30 [=======================>......] - ETA: 0s - loss: 0.3680 - accuracy: 0.8361

26/30 [=========================>....] - ETA: 0s - loss: 0.3686 - accuracy: 0.8366

28/30 [===========================>..] - ETA: 0s - loss: 0.3662 - accuracy: 0.8375

30/30 [==============================] - 3s 92ms/step - loss: 0.3667 - accuracy: 0.8364 - val_loss: 0.3885 - val_accuracy: 0.8070


Epoch 9/20


 1/30 [>.............................] - ETA: 9s - loss: 0.3447 - accuracy: 0.8457

 2/30 [=>............................] - ETA: 5s - loss: 0.3505 - accuracy: 0.8438

 3/30 [==>...........................] - ETA: 4s - loss: 0.3442 - accuracy: 0.8516

 5/30 [====>.........................] - ETA: 2s - loss: 0.3446 - accuracy: 0.8527

 7/30 [======>.......................] - ETA: 2s - loss: 0.3397 - accuracy: 0.8594

 8/30 [=======>......................] - ETA: 1s - loss: 0.3415 - accuracy: 0.8591

 9/30 [========>.....................] - ETA: 1s - loss: 0.3422 - accuracy: 0.8574

11/30 [==========>...................] - ETA: 1s - loss: 0.3429 - accuracy: 0.8551

13/30 [============>.................] - ETA: 1s - loss: 0.3425 - accuracy: 0.8511

15/30 [==============>...............] - ETA: 1s - loss: 0.3403 - accuracy: 0.8535

17/30 [================>.............] - ETA: 0s - loss: 0.3386 - accuracy: 0.8535

19/30 [==================>...........] - ETA: 0s - loss: 0.3398 - accuracy: 0.8521

21/30 [====================>.........] - ETA: 0s - loss: 0.3381 - accuracy: 0.8530

23/30 [======================>.......] - ETA: 0s - loss: 0.3367 - accuracy: 0.8539

25/30 [========================>.....] - ETA: 0s - loss: 0.3359 - accuracy: 0.8541

27/30 [==========================>...] - ETA: 0s - loss: 0.3350 - accuracy: 0.8547

29/30 [============================>.] - ETA: 0s - loss: 0.3337 - accuracy: 0.8559

30/30 [==============================] - 3s 91ms/step - loss: 0.3338 - accuracy: 0.8560 - val_loss: 0.3669 - val_accuracy: 0.8331


Epoch 10/20


 1/30 [>.............................] - ETA: 9s - loss: 0.3542 - accuracy: 0.8340

 2/30 [=>............................] - ETA: 5s - loss: 0.3425 - accuracy: 0.8496

 3/30 [==>...........................] - ETA: 4s - loss: 0.3249 - accuracy: 0.8587

 5/30 [====>.........................] - ETA: 2s - loss: 0.3204 - accuracy: 0.8574

 7/30 [======>.......................] - ETA: 2s - loss: 0.3169 - accuracy: 0.8647

 8/30 [=======>......................] - ETA: 1s - loss: 0.3181 - accuracy: 0.8643

10/30 [=========>....................] - ETA: 1s - loss: 0.3193 - accuracy: 0.8652

12/30 [===========>..................] - ETA: 1s - loss: 0.3194 - accuracy: 0.8634

14/30 [=============>................] - ETA: 1s - loss: 0.3165 - accuracy: 0.8645

16/30 [===============>..............] - ETA: 0s - loss: 0.3154 - accuracy: 0.8651

18/30 [=================>............] - ETA: 0s - loss: 0.3134 - accuracy: 0.8664

20/30 [===================>..........] - ETA: 0s - loss: 0.3119 - accuracy: 0.8681

22/30 [=====================>........] - ETA: 0s - loss: 0.3116 - accuracy: 0.8699

24/30 [=======================>......] - ETA: 0s - loss: 0.3098 - accuracy: 0.8695

26/30 [=========================>....] - ETA: 0s - loss: 0.3086 - accuracy: 0.8705

28/30 [===========================>..] - ETA: 0s - loss: 0.3065 - accuracy: 0.8720

30/30 [==============================] - 3s 91ms/step - loss: 0.3066 - accuracy: 0.8723 - val_loss: 0.3492 - val_accuracy: 0.8398


Epoch 11/20


 1/30 [>.............................] - ETA: 9s - loss: 0.3027 - accuracy: 0.8691

 2/30 [=>............................] - ETA: 5s - loss: 0.2787 - accuracy: 0.8799

 3/30 [==>...........................] - ETA: 4s - loss: 0.2782 - accuracy: 0.8854

 4/30 [===>..........................] - ETA: 3s - loss: 0.2742 - accuracy: 0.8838

 6/30 [=====>........................] - ETA: 2s - loss: 0.2734 - accuracy: 0.8818

 8/30 [=======>......................] - ETA: 1s - loss: 0.2724 - accuracy: 0.8816

10/30 [=========>....................] - ETA: 1s - loss: 0.2724 - accuracy: 0.8842

12/30 [===========>..................] - ETA: 1s - loss: 0.2737 - accuracy: 0.8841

14/30 [=============>................] - ETA: 1s - loss: 0.2778 - accuracy: 0.8855

16/30 [===============>..............] - ETA: 0s - loss: 0.2803 - accuracy: 0.8839

18/30 [=================>............] - ETA: 0s - loss: 0.2827 - accuracy: 0.8813

20/30 [===================>..........] - ETA: 0s - loss: 0.2834 - accuracy: 0.8805

22/30 [=====================>........] - ETA: 0s - loss: 0.2825 - accuracy: 0.8813

24/30 [=======================>......] - ETA: 0s - loss: 0.2822 - accuracy: 0.8818

26/30 [=========================>....] - ETA: 0s - loss: 0.2821 - accuracy: 0.8820

28/30 [===========================>..] - ETA: 0s - loss: 0.2828 - accuracy: 0.8820

30/30 [==============================] - 3s 92ms/step - loss: 0.2837 - accuracy: 0.8823 - val_loss: 0.3352 - val_accuracy: 0.8450


Epoch 12/20


 1/30 [>.............................] - ETA: 10s - loss: 0.2780 - accuracy: 0.8848

 2/30 [=>............................] - ETA: 5s - loss: 0.2780 - accuracy: 0.8828 

 3/30 [==>...........................] - ETA: 4s - loss: 0.2630 - accuracy: 0.8971

 4/30 [===>..........................] - ETA: 3s - loss: 0.2656 - accuracy: 0.8911

 6/30 [=====>........................] - ETA: 2s - loss: 0.2658 - accuracy: 0.8958

 8/30 [=======>......................] - ETA: 2s - loss: 0.2714 - accuracy: 0.8914

 9/30 [========>.....................] - ETA: 1s - loss: 0.2686 - accuracy: 0.8926

11/30 [==========>...................] - ETA: 1s - loss: 0.2680 - accuracy: 0.8928

13/30 [============>.................] - ETA: 1s - loss: 0.2643 - accuracy: 0.8953

15/30 [==============>...............] - ETA: 1s - loss: 0.2640 - accuracy: 0.8945

17/30 [================>.............] - ETA: 0s - loss: 0.2644 - accuracy: 0.8932

19/30 [==================>...........] - ETA: 0s - loss: 0.2615 - accuracy: 0.8941

21/30 [====================>.........] - ETA: 0s - loss: 0.2613 - accuracy: 0.8943

23/30 [======================>.......] - ETA: 0s - loss: 0.2608 - accuracy: 0.8952

25/30 [========================>.....] - ETA: 0s - loss: 0.2611 - accuracy: 0.8952

27/30 [==========================>...] - ETA: 0s - loss: 0.2610 - accuracy: 0.8948

29/30 [============================>.] - ETA: 0s - loss: 0.2618 - accuracy: 0.8949

30/30 [==============================] - 3s 94ms/step - loss: 0.2611 - accuracy: 0.8950 - val_loss: 0.3243 - val_accuracy: 0.8510


Epoch 13/20


 1/30 [>.............................] - ETA: 9s - loss: 0.2514 - accuracy: 0.9062

 2/30 [=>............................] - ETA: 5s - loss: 0.2510 - accuracy: 0.8994

 3/30 [==>...........................] - ETA: 4s - loss: 0.2557 - accuracy: 0.8945

 5/30 [====>.........................] - ETA: 2s - loss: 0.2534 - accuracy: 0.8980

 7/30 [======>.......................] - ETA: 2s - loss: 0.2494 - accuracy: 0.9009

 8/30 [=======>......................] - ETA: 1s - loss: 0.2470 - accuracy: 0.9021

 9/30 [========>.....................] - ETA: 1s - loss: 0.2489 - accuracy: 0.8993

11/30 [==========>...................] - ETA: 1s - loss: 0.2481 - accuracy: 0.8995

13/30 [============>.................] - ETA: 1s - loss: 0.2490 - accuracy: 0.8996

15/30 [==============>...............] - ETA: 1s - loss: 0.2460 - accuracy: 0.9020

17/30 [================>.............] - ETA: 0s - loss: 0.2462 - accuracy: 0.9030

19/30 [==================>...........] - ETA: 0s - loss: 0.2446 - accuracy: 0.9034

21/30 [====================>.........] - ETA: 0s - loss: 0.2430 - accuracy: 0.9047

23/30 [======================>.......] - ETA: 0s - loss: 0.2435 - accuracy: 0.9048

25/30 [========================>.....] - ETA: 0s - loss: 0.2425 - accuracy: 0.9054

27/30 [==========================>...] - ETA: 0s - loss: 0.2420 - accuracy: 0.9056

29/30 [============================>.] - ETA: 0s - loss: 0.2427 - accuracy: 0.9055

30/30 [==============================] - 3s 91ms/step - loss: 0.2430 - accuracy: 0.9057 - val_loss: 0.3168 - val_accuracy: 0.8520


Epoch 14/20


 1/30 [>.............................] - ETA: 9s - loss: 0.2481 - accuracy: 0.8984

 2/30 [=>............................] - ETA: 5s - loss: 0.2603 - accuracy: 0.8906

 3/30 [==>...........................] - ETA: 4s - loss: 0.2514 - accuracy: 0.8919

 4/30 [===>..........................] - ETA: 3s - loss: 0.2466 - accuracy: 0.8945

 6/30 [=====>........................] - ETA: 2s - loss: 0.2474 - accuracy: 0.8981

 8/30 [=======>......................] - ETA: 1s - loss: 0.2439 - accuracy: 0.9011

 9/30 [========>.....................] - ETA: 1s - loss: 0.2432 - accuracy: 0.9028

11/30 [==========>...................] - ETA: 1s - loss: 0.2429 - accuracy: 0.9047

13/30 [============>.................] - ETA: 1s - loss: 0.2389 - accuracy: 0.9066

15/30 [==============>...............] - ETA: 1s - loss: 0.2376 - accuracy: 0.9061

17/30 [================>.............] - ETA: 0s - loss: 0.2376 - accuracy: 0.9050

19/30 [==================>...........] - ETA: 0s - loss: 0.2339 - accuracy: 0.9067

21/30 [====================>.........] - ETA: 0s - loss: 0.2332 - accuracy: 0.9084

23/30 [======================>.......] - ETA: 0s - loss: 0.2301 - accuracy: 0.9096

25/30 [========================>.....] - ETA: 0s - loss: 0.2299 - accuracy: 0.9098

27/30 [==========================>...] - ETA: 0s - loss: 0.2285 - accuracy: 0.9096

29/30 [============================>.] - ETA: 0s - loss: 0.2278 - accuracy: 0.9104

30/30 [==============================] - 3s 91ms/step - loss: 0.2269 - accuracy: 0.9107 - val_loss: 0.3092 - val_accuracy: 0.8589


Epoch 15/20


 1/30 [>.............................] - ETA: 9s - loss: 0.2291 - accuracy: 0.9141

 2/30 [=>............................] - ETA: 5s - loss: 0.2444 - accuracy: 0.9092

 3/30 [==>...........................] - ETA: 4s - loss: 0.2467 - accuracy: 0.9095

 5/30 [====>.........................] - ETA: 2s - loss: 0.2351 - accuracy: 0.9152

 7/30 [======>.......................] - ETA: 2s - loss: 0.2249 - accuracy: 0.9199

 8/30 [=======>......................] - ETA: 1s - loss: 0.2234 - accuracy: 0.9192

 9/30 [========>.....................] - ETA: 1s - loss: 0.2245 - accuracy: 0.9193

11/30 [==========>...................] - ETA: 1s - loss: 0.2217 - accuracy: 0.9201

13/30 [============>.................] - ETA: 1s - loss: 0.2187 - accuracy: 0.9205

15/30 [==============>...............] - ETA: 0s - loss: 0.2167 - accuracy: 0.9197

17/30 [================>.............] - ETA: 0s - loss: 0.2155 - accuracy: 0.9202

19/30 [==================>...........] - ETA: 0s - loss: 0.2145 - accuracy: 0.9203

21/30 [====================>.........] - ETA: 0s - loss: 0.2130 - accuracy: 0.9209

23/30 [======================>.......] - ETA: 0s - loss: 0.2127 - accuracy: 0.9207

25/30 [========================>.....] - ETA: 0s - loss: 0.2123 - accuracy: 0.9208

27/30 [==========================>...] - ETA: 0s - loss: 0.2102 - accuracy: 0.9220

29/30 [============================>.] - ETA: 0s - loss: 0.2104 - accuracy: 0.9218

30/30 [==============================] - 3s 91ms/step - loss: 0.2117 - accuracy: 0.9213 - val_loss: 0.3059 - val_accuracy: 0.8665


Epoch 16/20


 1/30 [>.............................] - ETA: 9s - loss: 0.2238 - accuracy: 0.9316

 2/30 [=>............................] - ETA: 5s - loss: 0.1992 - accuracy: 0.9395

 3/30 [==>...........................] - ETA: 4s - loss: 0.1944 - accuracy: 0.9414

 5/30 [====>.........................] - ETA: 2s - loss: 0.2004 - accuracy: 0.9375

 7/30 [======>.......................] - ETA: 2s - loss: 0.1987 - accuracy: 0.9350

 8/30 [=======>......................] - ETA: 1s - loss: 0.2012 - accuracy: 0.9316

 9/30 [========>.....................] - ETA: 1s - loss: 0.2007 - accuracy: 0.9303

11/30 [==========>...................] - ETA: 1s - loss: 0.2041 - accuracy: 0.9284

13/30 [============>.................] - ETA: 1s - loss: 0.2012 - accuracy: 0.9292

15/30 [==============>...............] - ETA: 0s - loss: 0.2003 - accuracy: 0.9289

17/30 [================>.............] - ETA: 0s - loss: 0.1991 - accuracy: 0.9289

19/30 [==================>...........] - ETA: 0s - loss: 0.1985 - accuracy: 0.9278

21/30 [====================>.........] - ETA: 0s - loss: 0.1985 - accuracy: 0.9272

23/30 [======================>.......] - ETA: 0s - loss: 0.1971 - accuracy: 0.9279

25/30 [========================>.....] - ETA: 0s - loss: 0.1970 - accuracy: 0.9283

27/30 [==========================>...] - ETA: 0s - loss: 0.1966 - accuracy: 0.9283

29/30 [============================>.] - ETA: 0s - loss: 0.1979 - accuracy: 0.9279

30/30 [==============================] - 3s 90ms/step - loss: 0.1977 - accuracy: 0.9279 - val_loss: 0.3002 - val_accuracy: 0.8637


Epoch 17/20


 1/30 [>.............................] - ETA: 9s - loss: 0.1891 - accuracy: 0.9258

 2/30 [=>............................] - ETA: 5s - loss: 0.1884 - accuracy: 0.9307

 3/30 [==>...........................] - ETA: 3s - loss: 0.1993 - accuracy: 0.9232

 5/30 [====>.........................] - ETA: 2s - loss: 0.1964 - accuracy: 0.9262

 7/30 [======>.......................] - ETA: 2s - loss: 0.1936 - accuracy: 0.9266

 8/30 [=======>......................] - ETA: 1s - loss: 0.1921 - accuracy: 0.9275

 9/30 [========>.....................] - ETA: 1s - loss: 0.1887 - accuracy: 0.9277

11/30 [==========>...................] - ETA: 1s - loss: 0.1876 - accuracy: 0.9297

13/30 [============>.................] - ETA: 1s - loss: 0.1867 - accuracy: 0.9298

15/30 [==============>...............] - ETA: 1s - loss: 0.1873 - accuracy: 0.9299

17/30 [================>.............] - ETA: 0s - loss: 0.1875 - accuracy: 0.9297

19/30 [==================>...........] - ETA: 0s - loss: 0.1878 - accuracy: 0.9302

21/30 [====================>.........] - ETA: 0s - loss: 0.1861 - accuracy: 0.9317

23/30 [======================>.......] - ETA: 0s - loss: 0.1867 - accuracy: 0.9315

25/30 [========================>.....] - ETA: 0s - loss: 0.1860 - accuracy: 0.9320

27/30 [==========================>...] - ETA: 0s - loss: 0.1848 - accuracy: 0.9325

29/30 [============================>.] - ETA: 0s - loss: 0.1850 - accuracy: 0.9324

30/30 [==============================] - 3s 93ms/step - loss: 0.1840 - accuracy: 0.9326 - val_loss: 0.2971 - val_accuracy: 0.8674


Epoch 18/20


 1/30 [>.............................] - ETA: 9s - loss: 0.1683 - accuracy: 0.9434

 2/30 [=>............................] - ETA: 5s - loss: 0.1773 - accuracy: 0.9404

 3/30 [==>...........................] - ETA: 4s - loss: 0.1845 - accuracy: 0.9395

 4/30 [===>..........................] - ETA: 3s - loss: 0.1825 - accuracy: 0.9453

 6/30 [=====>........................] - ETA: 2s - loss: 0.1775 - accuracy: 0.9434

 8/30 [=======>......................] - ETA: 1s - loss: 0.1760 - accuracy: 0.9429

 9/30 [========>.....................] - ETA: 1s - loss: 0.1776 - accuracy: 0.9410

11/30 [==========>...................] - ETA: 1s - loss: 0.1760 - accuracy: 0.9412

13/30 [============>.................] - ETA: 1s - loss: 0.1753 - accuracy: 0.9395

15/30 [==============>...............] - ETA: 1s - loss: 0.1748 - accuracy: 0.9396

17/30 [================>.............] - ETA: 0s - loss: 0.1735 - accuracy: 0.9409

19/30 [==================>...........] - ETA: 0s - loss: 0.1722 - accuracy: 0.9417

21/30 [====================>.........] - ETA: 0s - loss: 0.1724 - accuracy: 0.9405

23/30 [======================>.......] - ETA: 0s - loss: 0.1723 - accuracy: 0.9403

25/30 [========================>.....] - ETA: 0s - loss: 0.1713 - accuracy: 0.9403

27/30 [==========================>...] - ETA: 0s - loss: 0.1721 - accuracy: 0.9397

29/30 [============================>.] - ETA: 0s - loss: 0.1733 - accuracy: 0.9390

30/30 [==============================] - 3s 93ms/step - loss: 0.1723 - accuracy: 0.9393 - val_loss: 0.2952 - val_accuracy: 0.8680


Epoch 19/20


 1/30 [>.............................] - ETA: 9s - loss: 0.1653 - accuracy: 0.9551

 2/30 [=>............................] - ETA: 5s - loss: 0.1513 - accuracy: 0.9629

 3/30 [==>...........................] - ETA: 4s - loss: 0.1556 - accuracy: 0.9570

 5/30 [====>.........................] - ETA: 2s - loss: 0.1579 - accuracy: 0.9500

 7/30 [======>.......................] - ETA: 2s - loss: 0.1583 - accuracy: 0.9487

 8/30 [=======>......................] - ETA: 1s - loss: 0.1601 - accuracy: 0.9470

 9/30 [========>.....................] - ETA: 1s - loss: 0.1621 - accuracy: 0.9462

11/30 [==========>...................] - ETA: 1s - loss: 0.1651 - accuracy: 0.9458

13/30 [============>.................] - ETA: 1s - loss: 0.1647 - accuracy: 0.9446

15/30 [==============>...............] - ETA: 1s - loss: 0.1642 - accuracy: 0.9444

17/30 [================>.............] - ETA: 0s - loss: 0.1609 - accuracy: 0.9453

19/30 [==================>...........] - ETA: 0s - loss: 0.1629 - accuracy: 0.9437

21/30 [====================>.........] - ETA: 0s - loss: 0.1616 - accuracy: 0.9443

23/30 [======================>.......] - ETA: 0s - loss: 0.1640 - accuracy: 0.9431

25/30 [========================>.....] - ETA: 0s - loss: 0.1638 - accuracy: 0.9421

27/30 [==========================>...] - ETA: 0s - loss: 0.1634 - accuracy: 0.9424

29/30 [============================>.] - ETA: 0s - loss: 0.1620 - accuracy: 0.9434

30/30 [==============================] - 3s 92ms/step - loss: 0.1611 - accuracy: 0.9437 - val_loss: 0.2947 - val_accuracy: 0.8702


Epoch 20/20


 1/30 [>.............................] - ETA: 9s - loss: 0.1502 - accuracy: 0.9473

 2/30 [=>............................] - ETA: 5s - loss: 0.1515 - accuracy: 0.9521

 3/30 [==>...........................] - ETA: 4s - loss: 0.1549 - accuracy: 0.9505

 5/30 [====>.........................] - ETA: 2s - loss: 0.1528 - accuracy: 0.9516

 7/30 [======>.......................] - ETA: 2s - loss: 0.1539 - accuracy: 0.9509

 8/30 [=======>......................] - ETA: 1s - loss: 0.1527 - accuracy: 0.9509

 9/30 [========>.....................] - ETA: 1s - loss: 0.1548 - accuracy: 0.9499

11/30 [==========>...................] - ETA: 1s - loss: 0.1528 - accuracy: 0.9490

13/30 [============>.................] - ETA: 1s - loss: 0.1503 - accuracy: 0.9495

15/30 [==============>...............] - ETA: 1s - loss: 0.1493 - accuracy: 0.9493

17/30 [================>.............] - ETA: 0s - loss: 0.1500 - accuracy: 0.9498

19/30 [==================>...........] - ETA: 0s - loss: 0.1495 - accuracy: 0.9494

21/30 [====================>.........] - ETA: 0s - loss: 0.1488 - accuracy: 0.9506

23/30 [======================>.......] - ETA: 0s - loss: 0.1514 - accuracy: 0.9493

25/30 [========================>.....] - ETA: 0s - loss: 0.1515 - accuracy: 0.9488

27/30 [==========================>...] - ETA: 0s - loss: 0.1517 - accuracy: 0.9490

29/30 [============================>.] - ETA: 0s - loss: 0.1515 - accuracy: 0.9492

30/30 [==============================] - 3s 94ms/step - loss: 0.1517 - accuracy: 0.9493 - val_loss: 0.2951 - val_accuracy: 0.8726


## Evaluate the model

And let's see how the model performs. Two values will be returned. Loss (a number which represents our error, lower values are better), and accuracy.

In [11]:
results = model.evaluate(test_data.batch(512), verbose=2)

for name, value in zip(model.metrics_names, results):
  print("%s: %.3f" % (name, value))

loss: 0.315
accuracy: 0.862


This fairly naive approach achieves an accuracy of about 87%. With more advanced approaches, the model should get closer to 95%.

## Further reading

For a more general way to work with string inputs and for a more detailed analysis of the progress of accuracy and loss during training, take a look [here](https://www.tensorflow.org/tutorials/keras/basic_text_classification).